src/include/misc.h

```diff
@@ -276,3 +276,21 @@ union __wt_rand_state {
 		uint32_t w, z;
 	} x;
 };
+
+/*
+ * WT_TAILQ_SAFE_REMOVE --
+ *	Macro to safely walk a TAILQ where we're expecting some underlying
+ * function to remove elements from the list, but we don't want to stop on
+ * error, nor do we want an error to turn into an infinite loop. Used during
+ * shutdown, when we're shutting down various lists. Unlike TAILQ_FOREACH_SAFE,
+ * this macro works even when the next element gets removed along with the
+ * current one.
+ */
+#define	WT_TAILQ_SAFE_REMOVE(var, head, field, tvar)			\
+	for ((tvar) = NULL; ((var) = TAILQ_FIRST(head)) != NULL;	\
+	    (tvar) = (var))						\
+		if ((tvar) == (var)) {					\
+			/* Leak the structure. */			\
+			TAILQ_REMOVE(head, (var), field);		\
+			continue;					\
+		} else
```

Comment:
@sueloverso and @keithbostic I hear you about the dangling `else`.  I don't think it's actually more dangerous that a macro that just includes a `for` without a body, but it's certainly not obvious.

@sulabhM and I kicked this around before he pushed that change -- I could live with BEGIN / END macros, but having `BEGIN { ... } END` seems messy (the braces aren't necessary but code won't autoformat correctly without them).  That said, we have exactly one other case of this, `WT_INTL_FOREACH_BEGIN { ... } END`, and it hasn't caused trouble.

TL;DR we should go back to the way this was and I'll ignore the unnecessary braces.